File: apps/scraper/query/sources/scholarships.py
"""The ASU scholarship search, filtered by what the student says about themselves."""

from __future__ import annotations

from scraper.core.types import QueryParam, QuerySource
from scraper.query.params import choices, code, text, url

# Option values of the exposed filters on onsa.asu.edu/scholarships.
_CITIZENSHIP = {
    "us citizen": "75",
    "us permanent resident": "76",
    "daca/dreamer": "77",
    "international student": "78",
}

_APPLICANTS = {
    "first-year undergrad": "58",
    "second-year undergrad": "59",
    "third-year undergrad": "60",
    "fourth-year+ undergrad": "61",
    "graduate student": "62",
    "undergraduate alumni": "63",
    "graduate alumni": "64",
}

_FOCUS = {
    "business and entrepreneurship": "44",
    "creative and performing arts": "45",
    "environment and sustainability": "46",
    "health and medicine": "47",
    "humanities": "48",
    "international affairs": "49",
    "journalism and media": "50",
    "national security": "51",
    "public policy": "53",
    "public service": "52",
    "social justice": "54",
    "social science": "57",
    "stem": "55",
    "peace and conflict resolution": "56",
}


def to_url(params: dict[str, str]) -> str:
    """The scholarship search URL with every filter that was given."""
    return url(
        "https://onsa.asu.edu/scholarships",
        [
            ("combine", text(params, "keywords")),
            ("field_citizenship_status", code(params, "citizenship", _CITIZENSHIP)),
            ("field_eligible_applicants", code(params, "applicant", _APPLICANTS)),
            ("field_focus", code(params, "focus", _FOCUS)),
        ],
    )


QUERY = QuerySource(
    key="scholarships",
    description=(
        "Search ASU scholarships by keyword, citizenship, year of study and field of focus."
    ),
    params=(
        QueryParam("keywords", "Words in the scholarship name or description.", example="women"),
        QueryParam("citizenship", "Citizenship status.", choices=choices(_CITIZENSHIP)),
        QueryParam("applicant", "Year of study.", choices=choices(_APPLICANTS)),
        QueryParam("focus", "Field of focus.", choices=choices(_FOCUS)),
    ),
    to_url=to_url,
)

File: apps/engine/src/runtime/tools/knowledge/search/scholarships.rs
//! ASU scholarships by keyword, citizenship, year of study and field.

use super::{LiveSource, Param};

/// The ASU scholarship search.
pub struct Scholarships;

const CITIZENSHIP: &[&str] = &[
    "us citizen",
    "us permanent resident",
    "daca/dreamer",
    "international student",
];

const APPLICANTS: &[&str] = &[
    "first-year undergrad",
    "second-year undergrad",
    "third-year undergrad",
    "fourth-year+ undergrad",
    "graduate student",
    "undergraduate alumni",
    "graduate alumni",
];

const FOCUS: &[&str] = &[
    "business and entrepreneurship",
    "creative and performing arts",
    "environment and sustainability",
    "health and medicine",
    "humanities",
    "international affairs",
    "journalism and media",
    "national security",
    "public policy",
    "public service",
    "social justice",
    "social science",
    "stem",
    "peace and conflict resolution",
];

const PARAMS: &[Param] = &[
    Param::text("keywords", "Words in the scholarship name or description.").example("women"),
    Param::one_of("citizenship", "Citizenship status.", CITIZENSHIP),
    Param::one_of("applicant", "Year of study.", APPLICANTS),
    Param::one_of("focus", "Field of focus.", FOCUS),
];

impl LiveSource for Scholarships {
    fn key(&self) -> &'static str {
        "scholarships"
    }

    fn hint(&self) -> &'static str {
        "scholarships and who may apply"
    }

    fn label(&self) -> &'static str {
        "ASU Scholarship Search"
    }

    fn category(&self) -> &'static str {
        "scholarships"
    }

    fn params(&self) -> &'static [Param] {
        PARAMS
    }
}
